src/neo/SmartContract/Native/Oracle/OracleContract.cs

```diff
@@ -0,0 +1,176 @@
+                if (!list.Remove(response.Id)) throw new InvalidOperationException();
+                if (list.Count == 0) engine.Snapshot.Storages.Delete(key);
+
+                //Mint GAS for oracle nodes
+                ECPoint[] nodes = GetOracleNodes(engine.Snapshot);
+                int index = (int)(response.Id % (ulong)nodes.Length);
+                UInt160 account = Contract.CreateSignatureRedeemScript(nodes[index]).ToScriptHash();
+                GAS.Mint(engine, account, OracleRequestPrice);
+            }
+        }
+
+        [ContractMethod(OracleRequestPrice, CallFlags.AllowModifyStates)]
+        private void Request(ApplicationEngine engine, string url, string filter, string callback, StackItem userData, long gasForRepsonse)
+        {
+            //Check the arguments
+            if (Utility.StrictUTF8.GetByteCount(url) > MaxUrlLength
+                || (filter != null && Utility.StrictUTF8.GetByteCount(filter) > MaxFilterLength)
+                || Utility.StrictUTF8.GetByteCount(callback) > MaxCallbackLength
+                || gasForRepsonse <= 0)
+                throw new ArgumentException();
+
+            //Mint gas for the response
+            engine.AddGas(gasForRepsonse);
+            GAS.Mint(engine, Hash, gasForRepsonse);
+
+            //Increase the request id
+            StorageItem item_id = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_RequestId));
+            ulong id = BitConverter.ToUInt64(item_id.Value) + 1;
+            item_id.Value = BitConverter.GetBytes(id);
+
+            //Put the request to storage
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_Request).Add(item_id.Value), new StorageItem(new OracleRequest
+            {
+                OriginalTxid = GetOriginalTxid(engine),
+                GasForResponse = gasForRepsonse,
+                Url = url,
+                Filter = filter,
+                CallbackContract = engine.CallingScriptHash,
+                CallbackMethod = callback,
+                UserData = BinarySerializer.Serialize(userData, MaxUserDataLength)
+            }));
+
+            //Add the id to the IdList
+            engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_IdList).Add(GetUrlHash(url)), () => new StorageItem(new IdList())).GetInteroperable<IdList>().Add(id);
+        }
+
+        [ContractMethod(0, CallFlags.AllowModifyStates)]
+        private void SetOracleNodes(ApplicationEngine engine, ECPoint[] nodes)
+        {
+            if (!CheckCommittee(engine)) throw new InvalidOperationException();
+            NodeList list = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_NodeList)).GetInteroperable<NodeList>();
+            list.Clear();
+            list.AddRange(nodes);
+        }
+
+        [ContractMethod(0_01000000, CallFlags.AllowStates)]
+        private bool Verify(ApplicationEngine engine)
```

Comment:
We need to reorder methods of `OracleContract`.

Reply:
What do you mean?

Reply:
Reorder the methods of `OracleContract`.

Reply:
Why? They are ordered by the method name.

Reply:
It's reading habits, I personally like the following order:
```cs
Verify
Request
Finish
PostPersist
```